var/spack/repos/builtin/packages/py-fava/package.py

```diff
@@ -0,0 +1,44 @@
+# Copyright 2013-2021 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+from spack import *
+
+
+class PyFava(PythonPackage):
+    """Fava is a web interface for the double-entry bookkeeping software
+       Beancount with a focus on features and usability."""
+
+    homepage = "https://beancount.github.io/fava/"
+    pypi     = "fava/fava-1.18.tar.gz"
+
+    version('1.18',  sha256='21336b695708497e6f00cab77135b174c51feb2713b657e0e208282960885bf5')
+
+    # For some reason Fava adds a whole bunch of executables to
+    # its bin directory, and this causes clashes when loading
+    # the module.
+    extends('python', ignore='bin/(bean-*|chardetect|cheroot|flask|py*|treeify|upload-to-sheets)')
+
+    # Some of the dependencies are not listed as required at
+    # build or run time, but actually are.
+    # - py-setuptools
+    # - py-importlib
+    # - py-pytest
+    depends_on('python@3.6:',       type=('build', 'run'))
+    depends_on('py-setuptools',     type='build')
+    depends_on('py-setuptools-scm', type=('build'))
+
+    depends_on('py-babel@2.6.0',         type=('build', 'run'))
+    depends_on('py-beancount@2.3.0:',    type=('build', 'run'))
+    depends_on('py-cheroot',             type=('build', 'run'))
+    depends_on('py-click',               type=('build', 'run'))
+    depends_on('py-flask@0.10.1:',       type=('build', 'run'))
+    depends_on('py-flask-babel@1.0.0:',  type=('build', 'run'))
+    depends_on('py-importlib',           type=('build', 'run'))
+    depends_on('py-jinja2@2.10:',        type=('build', 'run'))
+    depends_on('py-markdown2@2.3.0:',    type=('build', 'run'))
+    depends_on('py-ply',                 type=('build', 'run'))
+    depends_on('py-pytest',              type=('build', 'run'))
```

Comment:
Is this actually needed at run-time, or is it only needed to run the unit tests? Spack doesn't run the unit tests.

Reply:
Yes, its needed at run time. If its not available then an exception is thrown. It could even be that `beancount` is the package that *really* requires this. I'll check this out.

Reply:
```
  File "<PATH_TO_SPACK>/opt/spack/darwin-catalina-ivybridge/apple-clang-12.0.0/py-fava-1.18-45rq2kvefxe7c2qdz436b3ff5dh43ic5/lib/python3.8/site-packages/fava/__init__.py", line 2, in <module>
    from pkg_resources import DistributionNotFound
  File "<PATH_TO_SPACK>/opt/spack/darwin-catalina-ivybridge/apple-clang-12.0.0/py-setuptools-50.3.2-kn5xp4y2idojgcxjfhpfmi7hf2hykgvz/lib/python3.8/site-packages/pkg_resources/__init__.py", line 3239, in <module>
    def _initialize_master_working_set():
  File "<PATH_TO_SPACK>/opt/spack/darwin-catalina-ivybridge/apple-clang-12.0.0/py-setuptools-50.3.2-kn5xp4y2idojgcxjfhpfmi7hf2hykgvz/lib/python3.8/site-packages/pkg_resources/__init__.py", line 3222, in _call_aside
    f(*args, **kwargs)
  File "<PATH_TO_SPACK>/opt/spack/darwin-catalina-ivybridge/apple-clang-12.0.0/py-setuptools-50.3.2-kn5xp4y2idojgcxjfhpfmi7hf2hykgvz/lib/python3.8/site-packages/pkg_resources/__init__.py", line 3251, in _initialize_master_working_set
    working_set = WorkingSet._build_master()
  File "<PATH_TO_SPACK>/opt/spack/darwin-catalina-ivybridge/apple-clang-12.0.0/py-setuptools-50.3.2-kn5xp4y2idojgcxjfhpfmi7hf2hykgvz/lib/python3.8/site-packages/pkg_resources/__init__.py", line 567, in _build_master
    ws.require(__requires__)
  File "<PATH_TO_SPACK>/opt/spack/darwin-catalina-ivybridge/apple-clang-12.0.0/py-setuptools-50.3.2-kn5xp4y2idojgcxjfhpfmi7hf2hykgvz/lib/python3.8/site-packages/pkg_resources/__init__.py", line 884, in require
    needed = self.resolve(parse_requirements(requirements))
  File "<PATH_TO_SPACK>/opt/spack/darwin-catalina-ivybridge/apple-clang-12.0.0/py-setuptools-50.3.2-kn5xp4y2idojgcxjfhpfmi7hf2hykgvz/lib/python3.8/site-packages/pkg_resources/__init__.py", line 770, in resolve
    raise DistributionNotFound(req, requirers)
pkg_resources.DistributionNotFound: The 'pytest' distribution was not found and is required by beancount
```

Reply:
I'm not at all familiar with python packaging, so I can't quite tell whether [this](https://github.com/beancount/beancount/blob/ef3f17a81ac379e9768e207aa65a6ed5116bd135/setup.py#L81-L82) indicates that `pytest` is actually a dependency of `beancount` or not. All of the beancount scripts that `import pytest` are prefixed with `test_`, and so seem to indicate that they're a part of the [dev requirements](https://github.com/beancount/beancount/blob/70a390fa208a3d2d40717b941d28b7272b1fe014/requirements_dev.txt#L4). With `pytest` not listed as a dependency, I'm able to run some of the `beancount` scripts (e.g. `bean-check`) without error which is what I would have expected. But then again I know that this `beancount` framework offers the possibility for a user to do unit tests on their own beancount scripts (see [here for an example](https://beancount.github.io/docs/importing_external_data.html#regression-testing-your-importers)). Its a little unclear to me as to what the best course of action is. Any suggestions?

Reply:
I have no idea how the `fava/__init__.py` file is supposed to work. Let's just leave the pytest dep in there for now. You can open an issue with the devs if you want to ask about these undocumented deps.

Reply:
Thanks, I think that this is the simplest course of action for now. I'm relieved to hear you say it :-D